### PR TITLE
CrossEncoderModule with rerank API

### PR DESCRIPTION
**Pull request by markstur**
This module is closely related to EmbeddingModule.

Cross-encoder models use Q and A pairs and are trained return a relevance score for rank(). The existing rerank APIs in EmbeddingModule had to encode Q and A separately and use cosine similarity as a score. So the API is the same, but the results are supposed to be better (and slower).

Cross-encoder models do not support returning embedding vectors or sentence-similarity.

Support for the existing tokenization and model_info endpoints was also added.

### CODE REVIEW

File: caikit_nlp/modules/text_embedding/crossencoder.py [hunk 1]
```diff
@@ -0,0 +1,807 @@
...
+        return tokenized
+
+    def predict(
+        self,
+        sentences: List[List[str]],
+        batch_size: int = 32,
+        show_progress_bar: bool = None,
+        num_workers: int = 0,
+        activation_fct=None,
+        apply_softmax=False,
```
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/crossencoder.py:683]: Nope.  Our linter/formatter insists on no spaces around keyword parameter equals which is a good thing.

The odd thing is that lint/fmt rules are the opposite when there is a type.

Fortunately tox takes care of this.

File: caikit_nlp/modules/text_embedding/crossencoder.py [hunk 2]
```diff
@@ -0,0 +1,807 @@
...
+            producer_id=self.PRODUCER_ID,
+            input_token_count=results.input_token_count,
+        )
+
+    @RerankTasks.taskmethod()
+    def run_rerank_queries(
+        self,
+        queries: List[str],
+        documents: List[JsonDict],
+        top_n: Optional[int] = None,
```
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/crossencoder.py:344]: Long story, but short version is some folks thought that had other implications and preferred to avoid it and go with top_n.  This is now in our rerank API that is shared with text-embedding models and cross-encoder models so changing it would not be great.

So today I have top_n for the API we expose but it becomes top_k as the familiar parameter name in CrossEncoder functions.  Sorry.  Is there a better thing to do here?

File: caikit_nlp/modules/text_embedding/crossencoder.py [hunk 3]
```diff
@@ -0,0 +1,807 @@
...
+            *texts,
+            return_attention_mask=True,  # Used for determining token count
+            return_token_type_ids=False,  # Needed for cross-encoders
+            return_overflowing_tokens=False,  # DO NOT USE overflow tokens break sentence batches
+            return_offsets_mapping=True,  # Used for truncation test
+            return_length=False,
+            return_tensors="pt",
+            truncation=True,
+            padding=True,
+            max_length=max_len,
```
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/crossencoder.py:576]: Good catch, but after all that "do not touch this" strategy, Nick suggested we should have separate tokenizers per thread anyway.  Following that advice here (because he's probably right) and now we no longer get already borrowed issues when changing a param on a thread-dedicated tokenizer and don't need to write another truncation strategy (pair truncation).

I tested this with a multi-threaded client using randomized truncation to make sure I see the error before going tokenizer-per-thread and don't see it after.  I don't think that test fits our unit tests(?).

File: caikit_nlp/modules/text_embedding/crossencoder.py [hunk 4]
```diff
@@ -0,0 +1,807 @@
...
+
+                if end_index < len(texts[end_typeid][text_pos]):
+                    ret.append(i)
+
+        return ret
+
+    def smart_batching_collate_text_only(
+        self, batch, truncate_input_tokens: Optional[int] = 0
+    ):
+        texts = [[] for _ in range(len(batch[0]))]
```
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/crossencoder.py:634]: your way look better and "looks" equivalent but it breaks the data fetcher.  Using range works.

File: caikit_nlp/modules/text_embedding/crossencoder.py [hunk 5]
```diff
@@ -0,0 +1,807 @@
...
+                batch_size=32,
+                truncate_input_tokens=truncate_input_tokens,
+            )
+            results.append(scores)
+            input_token_count += token_count
+
+        # Fixup result dicts
+        for r in results:
+            for x in r:
+                x["score"] = float(x["score"].item())
```
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/crossencoder.py:425]: Not a tensor. I'll add convert_to_numpy=True (the default, but for clarity).  So here I always get a np.float.   I can get rid of the redundant float().  This is also after  the other change Gaurav suggested already using float().item() during the convert_to_numpy.

File: caikit_nlp/modules/text_embedding/crossencoder.py [hunk 6]
```diff
@@ -0,0 +1,807 @@
...
+    Token,
+    TokenizationResults,
+)
+from caikit.interfaces.nlp.tasks import RerankTask, RerankTasks, TokenizationTask
+import alog
+
+# Local
+from caikit_nlp.modules.text_embedding.utils import env_val_to_bool
+
+logger = alog.use_channel("TXT_EMB")
```
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/crossencoder.py]: CROSS_ENCODER looks good.  done

File: caikit_nlp/modules/text_embedding/crossencoder.py [hunk 7]
```diff
@@ -0,0 +1,807 @@
...
+error = error_handler.get(logger)
+
+
+RT = TypeVar("RT")  # return type
+
+
+class RerankResultTuple(NamedTuple):
+    """Output of modified predict()"""
```
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/crossencoder.py]: fixed

File: caikit_nlp/modules/text_embedding/crossencoder.py [hunk 8]
```diff
@@ -0,0 +1,807 @@
...
+            try:
+                model = torch.compile(model, backend=backend, mode="max-autotune")
+            except Exception as e:  # pylint: disable=broad-exception-caught
+                # Not always supported (e.g. in a python version) so catch, log, proceed.
+                warn_msg = (
+                    f"PT2_COMPILE enabled, but continuing without torch.compile() "
+                    f"because it failed with exception: {e}"
+                )
+                logger.warning(warn_msg, exc_info=True)
+        return model
```
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/crossencoder.py]: I was hoping to leverage all this and move it to utils, but things have changed enough and we won't be able to test IPEX in the first phase so I removed it.  Perhaps future optimization.

File: caikit_nlp/modules/text_embedding/crossencoder.py [hunk 9]
```diff
@@ -0,0 +1,807 @@
...
+        )
+
+        if results.results:
+            return RerankResult(
+                result=results.results[0],
+                producer_id=self.PRODUCER_ID,
+                input_token_count=results.input_token_count,
+            )
+
+        RerankResult(
```
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/crossencoder.py]: Unreachable.  Removed.

File: caikit_nlp/modules/text_embedding/crossencoder.py [hunk 10]
```diff
@@ -0,0 +1,807 @@
...
+
+        doc_texts = [get_text(doc) for doc in documents]
+
+        input_token_count = 0
+        results = []
+        for query in queries:
+            scores, token_count = self.model.rank(
+                query=query,
+                documents=doc_texts,
+                top_k=top_n,
```
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/crossencoder.py:325]: There was debate a year ago.  Our external API should stay top_n to avoid breaking the rerank API.

File: caikit_nlp/modules/text_embedding/crossencoder.py [hunk 11]
```diff
@@ -0,0 +1,807 @@
...
+                return_documents=False,
+                batch_size=32,
```
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/crossencoder.py:327]: I'd like to make that configurable, but pretty sure everyone is using the default so far for embeddings.  Will see how soon I can add that (will do more tests first).

File: caikit_nlp/modules/text_embedding/crossencoder.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/crossencoder.py:337]: This is always a np.float (after Gaurav's other suggestion).  I added convert_to_numpy=True param which we were using by default but seeing that helps for clarity.

File: tests/modules/text_embedding/test_crossencoder.py [hunk 12]
```diff
@@ -0,0 +1,443 @@
...
+
+    match1 = rf"exceeds the maximum sequence length for this model \({model_max}\) for text at indexes: 1, 3."
+    with pytest.raises(ValueError, match=match1):
+        loaded_model.run_rerank_queries(
+            queries=queries, documents=docs, truncate_input_tokens=truncate_input_tokens
+        )
+
+
+@pytest.mark.parametrize("truncate_input_tokens", [-1, 99, 510, 511, 512])
+def test_too_many_tokens_with_truncation_working(truncate_input_tokens, loaded_model):
```
[the remark this answers is not in the document]
Comment (markstur) [reply tests/modules/text_embedding/test_crossencoder.py:408]: Yep.  Needed to push the PR before I could get to that.  Will add some confirming test.

File: caikit_nlp/modules/text_embedding/crossencoder.py [hunk 13]
```diff
@@ -0,0 +1,807 @@
...
+        sentences: List[List[str]],
+        batch_size: int = 32,
+        show_progress_bar: bool = None,
+        num_workers: int = 0,
+        activation_fct=None,
+        apply_softmax=False,
+        convert_to_numpy: bool = True,
+        convert_to_tensor: bool = False,
+        truncate_input_tokens: Optional[int] = 0,
+    ) -> Union[PredictResultTuple, List[float], np.ndarray, torch.Tensor]:
```
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/crossencoder.py]: fixed

File: caikit_nlp/modules/text_embedding/crossencoder.py [hunk 14]
```diff
@@ -0,0 +1,807 @@
...
+        return_documents: bool = False,
+        batch_size: int = 32,
+        show_progress_bar: bool = None,
+        num_workers: int = 0,
+        activation_fct=None,
+        apply_softmax=False,
+        convert_to_numpy: bool = True,
+        convert_to_tensor: bool = False,
+        truncate_input_tokens: Optional[int] = 0,
+    ) -> Union[RerankResultTuple, List[Dict]]:
```
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/crossencoder.py]: fixed

File: caikit_nlp/modules/text_embedding/crossencoder.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/crossencoder.py:481]: oops, I see the docstring comments now.   Fixed

File: caikit_nlp/modules/text_embedding/crossencoder.py [hunk 15]
```diff
@@ -0,0 +1,807 @@
...
+        """
+
+        config = ModuleConfig.load(model_path)
+        error.dir_check("<NLP13823362E>", config.model_path)
+
+        artifacts_path = config.get(cls._ARTIFACTS_PATH_KEY)
+        error.value_check(
+            "<NLP20896115E>",
+            artifacts_path,
+            ValueError(f"Model config missing '{cls._ARTIFACTS_PATH_KEY}'"),
```
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/crossencoder.py]: Thanks.  Fixed

File: caikit_nlp/modules/text_embedding/crossencoder.py [hunk 16]
```diff
@@ -0,0 +1,807 @@
...
+        if use_device == "mps":
+            return mps
+        return "inductor"  # default backend
+
+    @staticmethod
+    def _optimize(model, ipex, device, autocast, pt2_compile):
+        if ipex:
+            if autocast:  # IPEX performs best with autocast using bfloat16
+                model = ipex.optimize(
+                    model, dtype=torch.bfloat16, weights_prepack=False
```
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/crossencoder.py]: Took this out because we won't really properly test the ipex options for short-term cross-encoder needs.  But FYI, various config names related to dtype/bfloat16 were rejected as confusing with other uses so for embeddings it ended up ipex + autocast is how you take advantage ipex with bloat16 speed.

File: caikit_nlp/modules/text_embedding/crossencoder.py [hunk 17]
```diff
@@ -0,0 +1,807 @@
...
+        )
+        inp_dataloader = DataLoader(
+            sentences,
+            batch_size=batch_size,
+            collate_fn=collate_fn,
+            num_workers=num_workers,
+            shuffle=False,
+        )
+
+        iterator = inp_dataloader
```
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/crossencoder.py]: done

File: caikit_nlp/modules/text_embedding/crossencoder.py [hunk 18]
```diff
@@ -0,0 +1,807 @@
...
+
+        if activation_fct is None:
+            activation_fct = self.default_activation_function
+
+        pred_scores = []
+        input_token_count = 0
+        self.model.eval()
+        self.model.to(self._target_device)
```
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/crossencoder.py]: makes sense.   Moved.

File: caikit_nlp/modules/text_embedding/crossencoder.py [hunk 19]
```diff
@@ -0,0 +1,807 @@
...
+                    input_token_count += sum(encoding.attention_mask)
+
+        if self.config.num_labels == 1:
+            pred_scores = [score[0] for score in pred_scores]
+
+        if convert_to_tensor:
+            pred_scores = torch.stack(pred_scores)
+        elif convert_to_numpy:
+            pred_scores = np.asarray(
+                [score.cpu().detach().numpy() for score in pred_scores]
```
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/crossencoder.py]: yes, your suggestion looks better.  Done

**Comment by markstur** [comment]
Thanks for the reviews!

Forgot to mention regarding the removal of ipex, etc code...  Part of that I was keeping to get MPS support as well, but I've found out that the default with CrossEncoder handles MPS and CUDA device already.

File: caikit_nlp/modules/text_embedding/crossencoder.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (gkumbhat) [reply caikit_nlp/modules/text_embedding/crossencoder.py:547]: interesting. Whats "data fetcher" here ?

File: caikit_nlp/modules/text_embedding/crossencoder.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (gkumbhat) [reply caikit_nlp/modules/text_embedding/crossencoder.py:596]: oh wow 🤔 thats weird

File: caikit_nlp/modules/text_embedding/crossencoder.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/crossencoder.py:547]: this is using torch DataLoader so callables and iterators or being used.   Overkill if I wrote it from scratch, but I'm using what sentence-transformers/CrossEncoder has been using as much as possible (with our extensions as needed for truncation and token counting).

File: tests/modules/text_embedding/test_crossencoder.py [hunk 20]
```diff
@@ -441,3 +447,68 @@ def test_too_many_tokens_with_truncation_working(truncate_input_tokens, loaded_m
...
+    for i, r in enumerate(queries_results):
+        queries_scores = [x.score for x in r.scores]
+        query_scores = [x.score for x in query_results[i].scores]
+        assert np.array_equal(queries_scores, query_scores)
+
+        # x...xyz is the same as x...xy because that is exactly where truncation worked
+        assert query_scores[0] == query_scores[1]
+
+        # Make sure the base, x, y are not a match (we kept the significant last char)
+        assert query_scores[1] != query_scores[2]
```
[the remark this answers is not in the document]
Comment (markstur) [reply tests/modules/text_embedding/test_crossencoder.py]: Yes. Unlike the embedding test I leveraged rerank() returns everything sorted by score instead of original index.  So I need to use the returned index to get them back into x, y, z order.

And then I got the == and != backwards (I guess to make it pass :( ) fixed that and tried to make that comment a little clearer

File: caikit_nlp/modules/text_embedding/crossencoder.py [hunk 21]
```diff
@@ -577,56 +494,42 @@ def get_tokenized(self, texts, **kwargs):
...
+
+        # Find the last offset by counting attn masks
+        # and keeping the last non-zero offset end.
+        token_count = 0
+        index = 0  # index of longest
+        type_id = 0  # track type_id of longest
+
+        for n, attn in enumerate(attn_mask):
+            if attn == 1:
+                token_count += 1
```
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/crossencoder.py:523]: it was just taking in one last (0, 0) that would never affect that last index that mapped a token back to the text.  So it seemed like the code was better off continuing, but I'll go look for unused stuff now.  Really seems like lint should catch that.

File: caikit_nlp/modules/text_embedding/crossencoder.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/crossencoder.py]: done

File: tests/modules/text_embedding/test_crossencoder.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (markstur) [reply tests/modules/text_embedding/test_crossencoder.py:414]: done
